### ISSUE TITLE
examples/tkvlc.py on MacOS with virtual environment fails due to sys.prefix is altered

### ISSUE THREAD

**Issue by kimwiktoren** [oaubert/python-vlc#131]
We only get audio due to not finding the libtk%.dylib library.

What I've discovered so far: when using:

`python -m venv venv`
`source venv/bin/activate`

sys.prefix is updated with the python path pointing to venv/lib.

By exchanging sys.prefix to the string '/Library/Frameworks/Python.framework/Versions/3.7' I manage to find the libtk%.dylib and load it with the current example.

My question is:
How can we find the correct path to the libtk%.dylib library when using virtual environments?

What I know:
There are some paths pointing towards the correct directory in sys.path, but they need to be parsed somehow.
There are the inspect module which can be used on any of the standard libraries that are imported + parsing to find the path to the /lib-folder.

These "solutions" feels sketchy, can it be solved in a more reliable way?

**Comment by mrJean1** [comment]
Question.  Python includes libtcl/-tk in its macOS distributions.  Why would those not be found the “normal” way in a `venv`?

**Comment by kimwiktoren** [comment]
The Python exemple use the prefix-variable, which aims at the virtual directory. But the standard libraries still exist in the default directory. Is the ”normal” way to concatenate prefix with lib?

**Comment by mrJean1** [comment]
Here is Python on plain macOS and inside a `venv`, before even running `tkvlc.py` ...

```
% python3.8
Python 3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18) 
...
>>> import Tkinter
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'Tkinter'
>>> import sys
>>> sys.prefix
'/Library/Frameworks/Python.framework/Versions/3.8'
```

```
% python3.7
Python 3.7.6 (v3.7.6:43364a7ae0, Dec 18 2019, 14:18:50) 
...
>>> import Tkinter
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'Tkinter'
>>> import sys
>>> sys.prefix
'/Library/Frameworks/Python.framework/Versions/3.7'
```

However for Python 2 ...

```
% python2.7
Python 2.7.18 (v2.7.18:8d21aa21f2, Apr 19 2020, 20:48:48) 
...
>>> import Tkinter
>>> Tkinter.__file__
'/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk/Tkinter.pyc'
>>> import sys
>>> sys.prefix
'/Library/Frameworks/Python.framework/Versions/2.7'
```

And in a `venv` ...

```
[~/Desktop/venv3] % python
Python 3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18) 
...
>>> import Tkinter
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'Tkinter'
>>> import sys
>>> sys.prefix
'~/Desktop/venv3'
>>>
```

That is one issue, how to get `Tkinter` inside a `venv`.  Perhaps, `tkvlc.py` should check `sys.prefix` plus the location of the Python executable, something like (to be investigated further):

```
>>> sys.executable
'~/Desktop/venv3/bin/python'
>>> import os
>>> os.__file__
'/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/os.py'
>>> os.path.dirname(os.__file__)
'/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8'
```

**Comment by oaubert** [comment]
On Debian and its derivatives (including Ubuntu), the tkinter module has been split out in its own package python3-tk, which is completely optional, and not installed by default by most distributions. It is the same for RedHat (python3-tkinter).

**Comment by mrJean1** [comment]
Python on macOS includes the Tcl/Tk libraries since recently, see [here](https://www.python.org/download/mac/tcltk/).  The 3.8 and 3.7 examples should have worked, but didn't.  Not sure yet why.

**Comment by mrJean1** [comment]
Here a correction of my previous post, using the proper module name.

```
% python3.8
Python 3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18) 
...
>>> import tkinter  # note tkinter, not Tkinter
>>> tkinter.__file__
'/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py'>>> import sys
>>> sys.prefix
'/Library/Frameworks/Python.framework/Versions/3.8'
```

```
% python3.7
Python 3.7.6 (v3.7.6:43364a7ae0, Dec 18 2019, 14:18:50) 
...
>>> import tkinter  # note tkinter, not Tkinter
>>> tkinter.__file__
'/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/tkinter/__init__.py'
>>> import sys
>>> sys.prefix
'/Library/Frameworks/Python.framework/Versions/3.7'
```

However for Python 2 ...

```
% python2.7
Python 2.7.18 (v2.7.18:8d21aa21f2, Apr 19 2020, 20:48:48) 
...
>>> import Tkinter  # note Tkinter, not tkinter
>>> Tkinter.__file__
'/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk/Tkinter.pyc'
>>> import sys
>>> sys.prefix
'/Library/Frameworks/Python.framework/Versions/2.7'
```

And in a `venv` ...

```
[~/Desktop/venv3] % python3
Python 3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18) 
...
>>> import tkinter  # note tkinter, not Tkinter
>>> tkinter.__file__
'/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tkinter/__init__.py'
>>> import sys
>>> sys.prefix
'~/Desktop/venv3'
>>>
```

Perhaps, the answer to the original issue is as follows.  If `tkinter.__file__` does not start with `sys.prefix`, use a portion of the former in lieu of the latter.  In `tkvlc.py`, this might look like

```
if _isMacOS:
    ...
    try:
        sys_prefix = sys.prefix
        if not Tk.__file__.startswith(sys_prefix):
            sys_prefix = Tk.__file__.split('/lib/python')[0]
        libtk = 'libtk%s.dylib' % (Tk.TkVersion,)
        libtk = joined(sys_prefix, 'lib', libtk)
```

Tested with Python 3.8.2, Tk 8.6, VLC 3.0.7 and `vlc.py` 3.0.7110, both in `venv` and plain macOS 10.13.6 High Sierra.

**Comment by mrJean1** [comment]
Yes, that is the “normal” way to find the C libraries in/from Python.  However, looking in the sys.prefix path in venv is not sufficient, clearly.

The example I just posted works, but is not the best.  Better would be to use the same mechanism Python uses to find its libraries when inside a venv.  But I’m not sure whether such a thing even exists.

/Jean

**Comment by kimwiktoren** [comment]
Thanks for looking into this! I saw your PR, even neater solution, even thou hidden.

**Comment by mrJean1** [comment]
Thank you for reporting the issue!

Those gymnastics are needed on macOS to find the TkMacOSXDrawableView for VLC and that in turn required getting the libtk.dylib.

There are still issues as described in the _Play method, but at least the media are shown now and … also in venv.

**Comment by kimwiktoren** [comment]
Yeah, been looking at that subwindows file as well. Seems like a Tkinter frame is always assumed to be top level? Annoying that the wrong view is returned... I did a test retrieving on both a parent frame and a child frame. Both returned the same view. Been "solving" it by doing a popup (TopLevel) window with the VLC-video instead as my whole app got covered.

`NSView *
TkMacOSXDrawableView(
    MacDrawable *macWin)
{
    NSView *result = nil;

    if (!macWin) {
	result = nil;
    } else if (!macWin->toplevel) {
	result = macWin->view;
    } else if (!(macWin->toplevel->flags & TK_EMBEDDED)) {
	result = macWin->toplevel->view;
    } else {
	TkWindow *contWinPtr = TkpGetOtherWindow(macWin->toplevel->winPtr);

	if (contWinPtr) {
	    result = TkMacOSXDrawableView(contWinPtr->privatePtr);
	}
    }
    return result;
}`

**Comment by mrJean1** [comment]
Interesting!  We might try to shoehorn this into tkvlc.py with ctypes or with some tcl.tk code.

**Comment by mrJean1** [comment]
kimwiktoren,

The current tkvlc.py version calls the TkMacOSXGetRootControl to get the drawable view, see the comments around tkvlc.py line 70

        _GetNSView = dylib.TkMacOSXGetRootControl

TkMacOSXGetRootControl is a public/global function accessable in the Tk library.  It calls private/local function TkMacOSXDrawableView and returns the NSView result of the latter.  In other words, tkvlc.py does get the drawable view you pointed out in your email below.

The issue is that vidoes playing in the TkMacOSXDrawableView overlap the buttons at the bottom of the window, see the comments at tkvlc.py line 323.

My question whether and if so how were you able to get around that problem?  By creating another, separate window, popup?  Please advise.

/Jean

**Comment by kimwiktoren** [comment]
mrJean1,

Yes, I'm aware of the GetRootControl. But I suspect that the function always return the actual root view of the whole app, not for the frame you send in. I.e. macWin->toplevel->view. This due to a small test I ran sending in the app frame and a child frame, comparing the _GetNSView return and they were referenced to the same object:

`
v1 = _GetNSView(app_frame)
v2 = _GetNSView(child_frame)
if v1 == v2:
   print('Same')
`

So my conclusion is that the video is drawn over the whole app due to this.

To get around it I started to play around with TopLevel windows. I got a working PoC with tkvlc.py where I create a TopLevel for the buttons, "anchoring" it to the bottom of the VLC-player. Looks pretty ok and the buttons are visible except in full screen mode.

Is this a solution you think would be acceptable?

<img width="500" alt="Screenshot 2020-05-06 at 19 54 04" src="https://user-images.githubusercontent.com/39512578/81211297-6f83c700-8fd3-11ea-89c2-a20058d60ab1.png">
